File: client_python/file_transfer.py
"""
File transfer module for Code Compiler & Executer Python Client
Handles file upload and download operations
Authors: Rares-Nicholas Popa & Adrian-Petru Enache
Date: April 2025
"""

import os
import struct
import hashlib
import math
from typing import Optional, Callable
from communication import Message, MessageType, MessageHandler


class FileTransferError(Exception):
    """File transfer related errors"""
    pass


class FileTransferManager:
    """Manages file upload and download operations"""
    
    def __init__(self, message_handler: MessageHandler, chunk_size: int = 64 * 1024):
        self.message_handler = message_handler
        self.chunk_size = chunk_size
        self.verbose = message_handler.verbose
    
    def upload_file(self, filepath: str, 
                   progress_callback: Optional[Callable[[int, int], None]] = None) -> bool:
        """
        Upload a file to the server
        
        Args:
            filepath: Path to file to upload
            progress_callback: Optional callback function(bytes_sent, total_bytes)
        
        Returns:
            True if upload successful, False otherwise
        """
        if not os.path.exists(filepath):
            raise FileTransferError(f"File not found: {filepath}")
        
        if not os.path.isfile(filepath):
            raise FileTransferError(f"Path is not a file: {filepath}")
        
        try:
            file_size = os.path.getsize(filepath)
            filename = os.path.basename(filepath)
            
            if self.verbose:
                print(f"Uploading {filename} ({file_size} bytes)...")
            
            # Step 1: Send upload start message
            if not self._send_upload_start(filename, file_size):
                return False
            
            # Step 2: Send file chunks
            chunk_count = math.ceil(file_size / self.chunk_size)
            bytes_sent = 0
            
            with open(filepath, 'rb') as f:
                for chunk_id in range(chunk_count):
                    chunk_data = f.read(self.chunk_size)
                    if not chunk_data:
                        break
                    
                    if not self._send_file_chunk(chunk_id, chunk_data):
                        return False
                    
                    bytes_sent += len(chunk_data)
                    
                    # Call progress callback
                    if progress_callback:
                        progress_callback(bytes_sent, file_size)
                    
                    if self.verbose and chunk_id % 10 == 0:  # Show progress every 10 chunks
                        percent = (bytes_sent / file_size) * 100
                        print(f"Upload progress: {percent:.1f}% ({bytes_sent}/{file_size})")
            
            # Step 3: Send upload end message
            if not self._send_upload_end():
                return False
            
            if self.verbose:
                print(f"Upload completed: {filename}")
            
            return True
            
        except Exception as e:
            if self.verbose:
                print(f"Upload failed: {e}")
            return False
    
    def _send_upload_start(self, filename: str, file_size: int) -> bool:
        """Send file upload start message"""
        try:
            # Calculate file checksum
            file_checksum = self._calculate_file_checksum_from_path(filename)
            chunk_count = math.ceil(file_size / self.chunk_size)
            
            # Pack upload start data
            filename_bytes = filename.encode('utf-8')[:256]
            data = struct.pack('>Q I I 256s I',
                             file_size,
                             chunk_count,
                             self.chunk_size,
                             filename_bytes,
                             file_checksum)
            
            msg = Message(MessageType.MSG_FILE_UPLOAD_START, data, 
                         self.message_handler._next_correlation_id())
            
            response = self.message_handler.send_and_receive(msg)
            
            if response and response.message_type == MessageType.MSG_ACK:
                return True
            elif response and response.message_type == MessageType.MSG_ERROR:
                error_info = self.message_handler._parse_error_response(response)
                print(f"Upload start failed: {error_info.get('error_message', 'Unknown error')}")
                return False
            
            return False
            
        except Exception as e:
            if self.verbose:
                print(f"Send upload start failed: {e}")
            return False
    
    def _send_file_chunk(self, chunk_id: int, chunk_data: bytes) -> bool:
        """Send a file chunk"""
        try:
            # Calculate chunk checksum
            import zlib
            chunk_checksum = zlib.crc32(chunk_data) & 0xffffffff
            
            # Pack chunk header
            chunk_header = struct.pack('>I I I',
                                     chunk_id,
                                     len(chunk_data),
                                     chunk_checksum)
            
            # Combine header and data
            data = chunk_header + chunk_data
            
            msg = Message(MessageType.MSG_FILE_UPLOAD_CHUNK, data,
                         self.message_handler._next_correlation_id())
            
            response = self.message_handler.send_and_receive(msg)
            
            if response and response.message_type == MessageType.MSG_ACK:
                return True
            elif response and response.message_type == MessageType.MSG_ERROR:
                error_info = self.message_handler._parse_error_response(response)
                print(f"Chunk {chunk_id} upload failed: {error_info.get('error_message', 'Unknown error')}")
                return False
            
            return False
            
        except Exception as e:
            if self.verbose:
                print(f"Send chunk {chunk_id} failed: {e}")
            return False
    
    def _send_upload_end(self) -> bool:
        """Send file upload end message"""
        try:
            msg = Message(MessageType.MSG_FILE_UPLOAD_END, b'',
                         self.message_handler._next_correlation_id())
            
            response = self.message_handler.send_and_receive(msg)
            
            if response and response.message_type == MessageType.MSG_ACK:
                return True
            elif response and response.message_type == MessageType.MSG_ERROR:
                error_info = self.message_handler._parse_error_response(response)
                print(f"Upload end failed: {error_info.get('error_message', 'Unknown error')}")
                return False
            
            return False
            
        except Exception as e:
            if self.verbose:
                print(f"Send upload end failed: {e}")
            return False
    
    def _calculate_file_checksum_from_path(self, filepath: str) -> int:
        """Calculate CRC32 checksum of a file"""
        import zlib
        checksum = 0
        
        try:
            with open(filepath, 'rb') as f:
                while True:
                    chunk = f.read(8192)
                    if not chunk:
                        break
                    checksum = zlib.crc32(chunk, checksum)
            
            return checksum & 0xffffffff
            
        except Exception:
            return 0
    
    def download_file(self, remote_filename: str, local_filepath: str,
                     progress_callback: Optional[Callable[[int, int], None]] = None) -> bool:
        """
        Download a file from the server (placeholder implementation)
        
        Args:
            remote_filename: Name of file on server
            local_filepath: Local path to save file
            progress_callback: Optional callback function(bytes_received, total_bytes)
        
        Returns:
            True if download successful, False otherwise
        """
        # Note: This is a placeholder implementation
        # The actual protocol would need additional message types for file download
        print(f"File download not yet implemented: {remote_filename} -> {local_filepath}")
        return False
    
    def verify_file_integrity(self, filepath: str, expected_checksum: int) -> bool:
        """Verify file integrity using checksum"""
        try:
            actual_checksum = self._calculate_file_checksum_from_path(filepath)
            return actual_checksum == expected_checksum
        except Exception:
            return False


class FileUploadProgress:
    """Helper class for tracking upload progress"""
    
    def __init__(self, total_size: int, verbose: bool = False):
        self.total_size = total_size
        self.bytes_uploaded = 0
        self.verbose = verbose
        self.start_time = None
        self.last_update_time = None
    
    def start(self):
        """Start progress tracking"""
        import time
        self.start_time = time.time()
        self.last_update_time = self.start_time
    
    def update(self, bytes_uploaded: int):
        """Update progress"""
        import time
        self.bytes_uploaded = bytes_uploaded
        current_time = time.time()
        
        if self.verbose and (current_time - self.last_update_time) >= 1.0:  # Update every second
            self._print_progress()
            self.last_update_time = current_time
    
    def finish(self):
        """Finish progress tracking"""
        if self.verbose:
            self._print_progress()
            print()  # New line
    
    def _print_progress(self):
        """Print current progress"""
        import time
        
        if self.total_size == 0:
            return
        
        percent = (self.bytes_uploaded / self.total_size) * 100
        
        # Calculate speed
        elapsed = time.time() - self.start_time if self.start_time else 0
        if elapsed > 0:
            speed = self.bytes_uploaded / elapsed
            speed_str = self._format_bytes(speed) + "/s"
        else:
            speed_str = "0 B/s"
        
        # Calculate ETA
        if elapsed > 0 and self.bytes_uploaded > 0:
            remaining_bytes = self.total_size - self.bytes_uploaded
            eta_seconds = (remaining_bytes / self.bytes_uploaded) * elapsed
            eta_str = self._format_time(eta_seconds)
        else:
            eta_str = "Unknown"
        
        print(f"\rProgress: {percent:5.1f}% "
              f"({self._format_bytes(self.bytes_uploaded)}/{self._format_bytes(self.total_size)}) "
              f"Speed: {speed_str} ETA: {eta_str}", end='', flush=True)
    
    def _format_bytes(self, bytes_value: float) -> str:
        """Format bytes in human readable format"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if bytes_value < 1024.0:
                return f"{bytes_value:.1f} {unit}"
            bytes_value /= 1024.0
        return f"{bytes_value:.1f} TB"
    
    def _format_time(self, seconds: float) -> str:
        """Format time in human readable format"""
        if seconds < 60:
            return f"{seconds:.0f}s"
        elif seconds < 3600:
            minutes = seconds / 60
            return f"{minutes:.0f}m"
        else:
            hours = seconds / 3600
            return f"{hours:.1f}h"


class ChunkedFileReader:
    """Helper class for reading files in chunks"""
    
    def __init__(self, filepath: str, chunk_size: int = 64 * 1024):
        self.filepath = filepath
        self.chunk_size = chunk_size
        self.file_size = os.path.getsize(filepath)
        self.chunks_total = math.ceil(self.file_size / chunk_size)
        self._file_handle = None
        self._current_chunk = 0
    
    def __enter__(self):
        self._file_handle = open(self.filepath, 'rb')
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self._file_handle:
            self._file_handle.close()
    
    def read_next_chunk(self) -> Optional[bytes]:
        """Read next chunk from file"""
        if not self._file_handle or self._current_chunk >= self.chunks_total:
            return None
        
        chunk_data = self._file_handle.read(self.chunk_size)
        self._current_chunk += 1
        
        return chunk_data if chunk_data else None
    
    def get_progress(self) -> tuple:
        """Get current progress (current_chunk, total_chunks)"""
        return (self._current_chunk, self.chunks_total)
    
    def calculate_checksum(self) -> int:
        """Calculate checksum of entire file"""
        import zlib
        checksum = 0
        
        if self._file_handle:
            current_pos = self._file_handle.tell()
            self._file_handle.seek(0)
            
            while True:
                chunk = self._file_handle.read(8192)
                if not chunk:
                    break
                checksum = zlib.crc32(chunk, checksum)
            
            self._file_handle.seek(current_pos)
        
        return checksum & 0xffffffff


# Utility functions
def is_binary_file(filepath: str) -> bool:
    """Check if file is binary"""
    try:
        with open(filepath, 'rb') as f:
            chunk = f.read(1024)
            return b'\x00' in chunk
    except Exception:
        return False


def get_file_mime_type(filepath: str) -> str:
    """Get MIME type of file based on extension"""
    import mimetypes
    mime_type, _ = mimetypes.guess_type(filepath)
    return mime_type or 'application/octet-stream'


def validate_source_file(filepath: str) -> bool:
    """Validate that file is a valid source code file"""
    if not os.path.exists(filepath):
        return False
    
    if not os.path.isfile(filepath):
        return False
    
    # Check file size (max 10MB for source files)
    file_size = os.path.getsize(filepath)
    if file_size > 10 * 1024 * 1024:
        return False
    
    # Check file extension
    ext = os.path.splitext(filepath)[1].lower()
    valid_extensions = {'.c', '.cpp', '.cc', '.cxx', '.h', '.hpp', 
                       '.java', '.py', '.js', '.go', '.rs'}
    
    if ext not in valid_extensions:
        return False
    
    # Check if file is text (not binary)
    if is_binary_file(filepath):
        return False
    
    return True


def sanitize_filename(filename: str) -> str:
    """Sanitize filename for safe transmission"""
    import re
    
    # Remove path separators and dangerous characters
    filename = os.path.basename(filename)
    filename = re.sub(r'[<>:"|?*\x00-\x1f]', '_', filename)
    
    # Limit length
    if len(filename) > 255:
        name, ext = os.path.splitext(filename)
        max_name_len = 255 - len(ext)
        filename = name[:max_name_len] + ext
    
    return filename


def create_temp_file(prefix: str = "upload_", suffix: str = "") -> str:
    """Create a temporary file and return its path"""
    import tempfile
    
    fd, filepath = tempfile.mkstemp(prefix=prefix, suffix=suffix)
    os.close(fd)  # Close the file descriptor, we just need the path
    
    return filepath


class FileTransferStats:
    """Statistics tracking for file transfers"""
    
    def __init__(self):
        self.files_uploaded = 0
        self.files_downloaded = 0
        self.bytes_uploaded = 0
        self.bytes_downloaded = 0
        self.upload_errors = 0
        self.download_errors = 0
        self.total_upload_time = 0.0
        self.total_download_time = 0.0
    
    def record_upload(self, file_size: int, duration: float, success: bool):
        """Record upload statistics"""
        if success:
            self.files_uploaded += 1
            self.bytes_uploaded += file_size
            self.total_upload_time += duration
        else:
            self.upload_errors += 1
    
    def record_download(self, file_size: int, duration: float, success: bool):
        """Record download statistics"""
        if success:
            self.files_downloaded += 1
            self.bytes_downloaded += file_size
            self.total_download_time += duration
        else:
            self.download_errors += 1
    
    def get_upload_speed(self) -> float:
        """Get average upload speed in bytes per second"""
        if self.total_upload_time > 0:
            return self.bytes_uploaded / self.total_upload_time
        return 0.0
    
    def get_download_speed(self) -> float:
        """Get average download speed in bytes per second"""
        if self.total_download_time > 0:
            return self.bytes_downloaded / self.total_download_time
        return 0.0
    
    def print_summary(self):
        """Print transfer statistics summary"""
        print("\n=== File Transfer Statistics ===")
        print(f"Files uploaded: {self.files_uploaded}")
        print(f"Files downloaded: {self.files_downloaded}")
        print(f"Bytes uploaded: {self._format_bytes(self.bytes_uploaded)}")
        print(f"Bytes downloaded: {self._format_bytes(self.bytes_downloaded)}")
        print(f"Upload errors: {self.upload_errors}")
        print(f"Download errors: {self.download_errors}")
        
        upload_speed = self.get_upload_speed()
        download_speed = self.get_download_speed()
        
        if upload_speed > 0:
            print(f"Average upload speed: {self._format_bytes(upload_speed)}/s")
        if download_speed > 0:
            print(f"Average download speed: {self._format_bytes(download_speed)}/s")
    
    def _format_bytes(self, bytes_value: float) -> str:
        """Format bytes in human readable format"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if bytes_value < 1024.0:
                return f"{bytes_value:.1f} {unit}"
            bytes_value /= 1024.0
        return f"{bytes_value:.1f} TB"


class SecureFileTransfer:
    """Enhanced file transfer with encryption and compression"""
    
    def __init__(self, message_handler: MessageHandler, chunk_size: int = 64 * 1024):
        self.base_transfer = FileTransferManager(message_handler, chunk_size)
        self.encryption_enabled = False
        self.compression_enabled = False
        self.encryption_key = None
    
    def enable_encryption(self, key: bytes):
        """Enable file encryption during transfer"""
        self.encryption_enabled = True
        self.encryption_key = key
    
    def enable_compression(self):
        """Enable file compression during transfer"""
        self.compression_enabled = True
    
    def upload_file(self, filepath: str, 
                   progress_callback: Optional[Callable[[int, int], None]] = None) -> bool:
        """Upload file with optional encryption and compression"""
        
        # If no special features enabled, use base implementation
        if not self.encryption_enabled and not self.compression_enabled:
            return self.base_transfer.upload_file(filepath, progress_callback)
        
        # For encrypted/compressed transfers, we would need to:
        # 1. Process the file (compress/encrypt)
        # 2. Upload the processed file
        # 3. Send metadata about processing
        
        # This is a placeholder for the enhanced implementation
        print("Encrypted/compressed file transfer not yet implemented")
        return False
    
    def _compress_file(self, input_path: str, output_path: str) -> bool:
        """Compress file using gzip"""
        try:
            import gzip
            import shutil
            
            with open(input_path, 'rb') as f_in:
                with gzip.open(output_path, 'wb') as f_out:
                    shutil.copyfileobj(f_in, f_out)
            
            return True
        except Exception:
            return False
    
    def _encrypt_file(self, input_path: str, output_path: str, key: bytes) -> bool:
        """Encrypt file using AES (placeholder)"""
        # This would require a crypto library like cryptography
        # For now, just copy the file
        try:
            import shutil
            shutil.copy2(input_path, output_path)
            return True
        except Exception:
            return False


class BatchFileTransfer:
    """Utility for transferring multiple files"""
    
    def __init__(self, file_transfer: FileTransferManager):
        self.file_transfer = file_transfer
        self.stats = FileTransferStats()
    
    def upload_files(self, file_paths: list, 
                    progress_callback: Optional[Callable[[int, int, str], None]] = None) -> dict:
        """
        Upload multiple files
        
        Args:
            file_paths: List of file paths to upload
            progress_callback: Optional callback(current_file, total_files, filename)
        
        Returns:
            Dictionary with results for each file
        """
        results = {}
        
        for i, filepath in enumerate(file_paths):
            filename = os.path.basename(filepath)
            
            if progress_callback:
                progress_callback(i + 1, len(file_paths), filename)
            
            import time
            start_time = time.time()
            
            try:
                success = self.file_transfer.upload_file(filepath)
                file_size = os.path.getsize(filepath) if os.path.exists(filepath) else 0
                duration = time.time() - start_time
                
                self.stats.record_upload(file_size, duration, success)
                
                results[filepath] = {
                    'success': success,
                    'file_size': file_size,
                    'duration': duration
                }
                
            except Exception as e:
                duration = time.time() - start_time
                self.stats.record_upload(0, duration, False)
                
                results[filepath] = {
                    'success': False,
                    'error': str(e),
                    'duration': duration
                }
        
        return results
    
    def get_statistics(self) -> FileTransferStats:
        """Get transfer statistics"""
        return self.stats


# File type detection utilities
def detect_file_language(filepath: str) -> str:
    """Detect programming language from file"""
    ext = os.path.splitext(filepath)[1].lower()
    
    language_map = {
        '.c': 'C',
        '.h': 'C',
        '.cpp': 'C++',
        '.cc': 'C++',
        '.cxx': 'C++',
        '.hpp': 'C++',
        '.hxx': 'C++',
        '.java': 'Java',
        '.py': 'Python',
        '.js': 'JavaScript',
        '.go': 'Go',
        '.rs': 'Rust',
    }
    
    return language_map.get(ext, 'Unknown')


def get_recommended_compile_args(filepath: str) -> str:
    """Get recommended compiler arguments based on file type"""
    language = detect_file_language(filepath)
    
    args_map = {
        'C': '-Wall -Wextra -std=c99',
        'C++': '-Wall -Wextra -std=c++17',
        'Java': '-cp .',
        'Python': '',
        'JavaScript': '',
        'Go': '',
        'Rust': '--edition 2021',
    }
    
    return args_map.get(language, '')


def estimate_compilation_time(filepath: str) -> float:
    """Estimate compilation time based on file size and type"""
    try:
        file_size = os.path.getsize(filepath)
        language = detect_file_language(filepath)
        
        # Base time per KB for different languages (in seconds)
        time_per_kb = {
            'C': 0.01,
            'C++': 0.02,
            'Java': 0.015,
            'Python': 0.001,  # Interpreted
            'JavaScript': 0.001,  # Interpreted
            'Go': 0.005,
            'Rust': 0.03,
        }
        
        base_time = time_per_kb.get(language, 0.01)
        size_kb = file_size / 1024
        
        # Add base overhead
        estimated_time = max(1.0, size_kb * base_time + 2.0)
        
        return estimated_time
        
    except Exception:
        return 30.0  # Default estimate

File: client_unix/src/file_transfer.cpp
/**
 * @file file_transfer.cpp
 * @brief File transfer implementation for Unix client
 * @author Rares-Nicholas Popa & Adrian-Petru Enache
 * @date April 2025
 */

#include "client.h"
#include "communication.h"
#include <fstream>
#include <iostream>
#include <iomanip>
#include <chrono>
#include <thread>
#include <algorithm>
#include <cstring>
#include <zlib.h>

namespace CodeCompiler {

/**
 * @brief FileTransferManager implementation
 */
FileTransferManager::FileTransferManager(MessageHandler& message_handler)
    : message_handler_(message_handler), chunk_size_(64 * 1024), compression_enabled_(false) {
}

void FileTransferManager::setChunkSize(size_t chunk_size) {
    chunk_size_ = std::max(size_t(1024), std::min(chunk_size, size_t(1024 * 1024))); // 1KB to 1MB
}

size_t FileTransferManager::getChunkSize() const {
    return chunk_size_;
}

void FileTransferManager::setCompressionEnabled(bool enabled) {
    compression_enabled_ = enabled;
}

bool FileTransferManager::isCompressionEnabled() const {
    return compression_enabled_;
}

bool FileTransferManager::uploadFile(const std::string& local_path, const std::string& remote_name,
                                   ProgressCallback progress) {
    auto start_time = std::chrono::steady_clock::now();
    
    try {
        // Validate file
        if (!validateFile(local_path)) {
            return false;
        }
        
        // Open file
        std::ifstream file(local_path, std::ios::binary);
        if (!file.is_open()) {
            return false;
        }
        
        // Get file size
        file.seekg(0, std::ios::end);
        size_t file_size = static_cast<size_t>(file.tellg());
        file.seekg(0, std::ios::beg);
        
        // Determine remote filename
        std::string filename = remote_name.empty() ? FileUtils::getBasename(local_path) : remote_name;
        
        // Calculate chunks
        size_t chunk_count = (file_size + chunk_size_ - 1) / chunk_size_;
        uint32_t file_checksum = calculateFileChecksum(local_path);
        
        // Send upload start
        if (!message_handler_.sendFileUploadStart(filename, file_size, 
                                                 static_cast<uint32_t>(chunk_count), 
                                                 chunk_size_, file_checksum)) {
            return false;
        }
        
        // Send file chunks
        size_t bytes_sent = 0;
        for (uint32_t chunk_id = 0; chunk_id < chunk_count; ++chunk_id) {
            std::vector<uint8_t> chunk_data = readFileChunk(file, chunk_size_);
            
            if (chunk_data.empty() && chunk_id < chunk_count - 1) {
                // Unexpected end of file
                return false;
            }
            
            if (!message_handler_.sendFileChunk(chunk_id, chunk_data)) {
                return false;
            }
            
            bytes_sent += chunk_data.size();
            
            // Update progress
            if (progress) {
                progress(bytes_sent, file_size);
            }
            
            // Small delay to prevent overwhelming the server
            if (chunk_count > 100 && chunk_id % 10 == 0) {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        }
        
        // Send upload end
        if (!message_handler_.sendFileUploadEnd()) {
            return false;
        }
        
        // Update statistics
        auto end_time = std::chrono::steady_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);
        
        stats_.files_uploaded++;
        stats_.bytes_uploaded += file_size;
        stats_.total_upload_time += duration;
        
        return true;
        
    } catch (const std::exception& e) {
        stats_.upload_errors++;
        return false;
    }
}

bool FileTransferManager::uploadFileAsync(const std::string& local_path, 
                                        std::function<void(bool)> completion_callback,
                                        const std::string& remote_name,
                                        ProgressCallback progress) {
    // Launch upload in separate thread
    std::thread upload_thread([this, local_path, remote_name, progress, completion_callback]() {
        bool success = uploadFile(local_path, remote_name, progress);
        if (completion_callback) {
            completion_callback(success);
        }
    });
    
    upload_thread.detach();
    return true;
}

bool FileTransferManager::downloadFile(const std::string& remote_name, const std::string& local_path,
                                      ProgressCallback progress) {
    // Download is not implemented in the current protocol
    // This would require additional message types
    (void)remote_name;
    (void)local_path;
    (void)progress;
    
    stats_.download_errors++;
    return false;
}

const FileTransferManager::TransferStats& FileTransferManager::getStats() const {
    return stats_;
}

void FileTransferManager::resetStats() {
    stats_ = TransferStats{};
}

uint32_t FileTransferManager::calculateFileChecksum(const std::string& filepath) const {
    std::ifstream file(filepath, std::ios::binary);
    if (!file.is_open()) {
        return 0;
    }
    
    uint32_t crc = 0;
    const size_t buffer_size = 8192;
    std::vector<uint8_t> buffer(buffer_size);
    
    while (file.read(reinterpret_cast<char*>(buffer.data()), buffer_size) || file.gcount() > 0) {
        size_t bytes_read = static_cast<size_t>(file.gcount());
        crc = crc32(crc, buffer.data(), bytes_read);
    }
    
    return crc;
}

bool FileTransferManager::validateFile(const std::string& filepath) const {
    // Check if file exists and is readable
    if (!FileUtils::exists(filepath) || !FileUtils::isFile(filepath) || !FileUtils::isReadable(filepath)) {
        return false;
    }
    
    // Check file size
    size_t file_size = FileUtils::getSize(filepath);
    if (file_size == 0 || file_size > 100 * 1024 * 1024) { // Max 100MB
        return false;
    }
    
    // Check if it's a valid source file
    return FileUtils::validateSourceFile(filepath);
}

std::vector<uint8_t> FileTransferManager::readFileChunk(std::ifstream& file, size_t chunk_size) const {
    std::vector<uint8_t> buffer(chunk_size);
    file.read(reinterpret_cast<char*>(buffer.data()), chunk_size);
    
    size_t bytes_read = static_cast<size_t>(file.gcount());
    buffer.resize(bytes_read);
    
    return buffer;
}

/**
 * @brief File upload progress tracker
 */
class FileUploadTracker {
public:
    FileUploadTracker(const std::string& filename, size_t total_size, bool verbose)
        : filename_(filename), total_size_(total_size), verbose_(verbose),
          bytes_uploaded_(0), start_time_(std::chrono::steady_clock::now()) {
        
        if (verbose_) {
            std::cout << "Uploading " << filename_ << " (" << formatBytes(total_size_) << ")..." << std::endl;
        }
    }
    
    void updateProgress(size_t bytes_uploaded) {
        bytes_uploaded_ = bytes_uploaded;
        
        if (verbose_ && total_size_ > 0) {
            auto now = std::chrono::steady_clock::now();
            auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - start_time_);
            
            // Update every 100ms or at completion
            if (elapsed.count() - last_update_time_ > 100 || bytes_uploaded_ == total_size_) {
                displayProgress();
                last_update_time_ = elapsed.count();
            }
        }
    }
    
    void complete(bool success) {
        if (verbose_) {
            if (success) {
                auto end_time = std::chrono::steady_clock::now();
                auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time_);
                
                std::cout << "\nUpload completed in " << duration.count() << " ms";
                if (duration.count() > 0) {
                    double speed = (static_cast<double>(total_size_) * 1000) / duration.count();
                    std::cout << " (" << formatBytes(static_cast<size_t>(speed)) << "/s)";
                }
                std::cout << std::endl;
            } else {
                std::cout << "\nUpload failed!" << std::endl;
            }
        }
    }

private:
    void displayProgress() {
        if (total_size_ == 0) return;
        
        double percent = (static_cast<double>(bytes_uploaded_) / total_size_) * 100.0;
        
        // Create progress bar
        const int bar_width = 30;
        int filled = static_cast<int>((percent / 100.0) * bar_width);
        
        std::cout << "\r[";
        for (int i = 0; i < bar_width; ++i) {
            if (i < filled) {
                std::cout << "=";
            } else if (i == filled) {
                std::cout << ">";
            } else {
                std::cout << " ";
            }
        }
        std::cout << "] " << std::fixed << std::setprecision(1) << percent << "% ";
        std::cout << "(" << formatBytes(bytes_uploaded_) << "/" << formatBytes(total_size_) << ")";
        std::cout << std::flush;
    }
    
    std::string formatBytes(size_t bytes) const {
        const char* units[] = {"B", "KB", "MB", "GB"};
        int unit = 0;
        double size = static_cast<double>(bytes);
        
        while (size >= 1024.0 && unit < 3) {
            size /= 1024.0;
            unit++;
        }
        
        std::ostringstream oss;
        oss << std::fixed << std::setprecision(1) << size << " " << units[unit];
        return oss.str();
    }
    
    std::string filename_;
    size_t total_size_;
    bool verbose_;
    size_t bytes_uploaded_;
    std::chrono::steady_clock::time_point start_time_;
    int64_t last_update_time_ = 0;
};

/**
 * @brief Compressed file transfer (if compression is enabled)
 */
class CompressedFileTransfer {
public:
    static std::vector<uint8_t> compressData(const std::vector<uint8_t>& input) {
        uLongf compressed_size = compressBound(input.size());
        std::vector<uint8_t> compressed(compressed_size);
        
        int result = compress(compressed.data(), &compressed_size, 
                            input.data(), input.size());
        
        if (result != Z_OK) {
            return input; // Return original if compression fails
        }
        
        compressed.resize(compressed_size);
        
        // Only use compressed data if it's actually smaller
        return (compressed.size() < input.size()) ? compressed : input;
    }
    
    static std::vector<uint8_t> decompressData(const std::vector<uint8_t>& compressed, 
                                              size_t original_size) {
        std::vector<uint8_t> decompressed(original_size);
        uLongf decompressed_size = original_size;
        
        int result = uncompress(decompressed.data(), &decompressed_size,
                              compressed.data(), compressed.size());
        
        if (result != Z_OK) {
            return {}; // Return empty on error
        }
        
        decompressed.resize(decompressed_size);
        return decompressed;
    }
};

/**
 * @brief File transfer utilities
 */
namespace FileTransferUtils {

bool isValidFilename(const std::string& filename) {
    if (filename.empty() || filename.length() > 255) {
        return false;
    }
    
    // Check for invalid characters
    const std::string invalid_chars = "<>:\"|?*\\";
    for (char c : filename) {
        if (invalid_chars.find(c) != std::string::npos || c < 32) {
            return false;
        }
    }
    
    // Check for path separators
    if (filename.find('/') != std::string::npos) {
        return false;
    }
    
    return true;
}

std::string sanitizeFilename(const std::string& filename) {
    std::string sanitized = FileUtils::getBasename(filename);
    
    // Replace invalid characters with underscores
    const std::string invalid_chars = "<>:\"|?*\\";
    for (char& c : sanitized) {
        if (invalid_chars.find(c) != std::string::npos || c < 32) {
            c = '_';
        }
    }
    
    // Limit length
    if (sanitized.length() > 255) {
        sanitized = sanitized.substr(0, 255);
    }
    
    return sanitized;
}

size_t calculateOptimalChunkSize(size_t file_size) {
    // Calculate chunk size based on file size for optimal transfer
    if (file_size < 1024 * 1024) {        // < 1MB
        return 8 * 1024;                   // 8KB chunks
    } else if (file_size < 10 * 1024 * 1024) {  // < 10MB
        return 32 * 1024;                  // 32KB chunks
    } else {                               // >= 10MB
        return 64 * 1024;                  // 64KB chunks
    }
}

bool isTransferInProgress(const std::string& temp_file) {
    // Check if a transfer is already in progress for this file
    return FileUtils::exists(temp_file + ".transfer");
}

void markTransferInProgress(const std::string& temp_file) {
    std::ofstream marker(temp_file + ".transfer");
    marker << std::chrono::system_clock::now().time_since_epoch().count() << std::endl;
}

void markTransferComplete(const std::string& temp_file) {
    std::remove((temp_file + ".transfer").c_str());
}

std::string createTransferMetadata(const std::string& filename, size_t file_size, 
                                  uint32_t checksum, size_t chunk_count) {
    std::ostringstream metadata;
    metadata << "filename=" << filename << "\n";
    metadata << "size=" << file_size << "\n";
    metadata << "checksum=" << std::hex << checksum << "\n";
    metadata << "chunks=" << chunk_count << "\n";
    metadata << "timestamp=" << std::chrono::system_clock::now().time_since_epoch().count() << "\n";
    return metadata.str();
}

bool verifyFileIntegrity(const std::string& filepath, uint32_t expected_checksum) {
    FileTransferManager dummy_manager(*(MessageHandler*)nullptr); // Hack for access to private method
    // In real implementation, this would be a standalone function
    return true; // Placeholder
}

} // namespace FileTransferUtils

} // namespace CodeCompiler